lib/observability/tracing/collector.go

```diff
@@ -0,0 +1,185 @@
+// Copyright 2022 Gravitational, Inc
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package tracing
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"io"
+	"net"
+	"net/http"
+
+	"github.com/gravitational/trace"
+	coltracepb "go.opentelemetry.io/proto/otlp/collector/trace/v1"
+	otlp "go.opentelemetry.io/proto/otlp/trace/v1"
+	"golang.org/x/sync/errgroup"
+	"google.golang.org/grpc"
+	"google.golang.org/grpc/credentials"
+	"google.golang.org/grpc/credentials/insecure"
+	"google.golang.org/protobuf/proto"
+)
+
+// Collector is a simple in memory implementation of an OpenTelemetry Collector
+// that is used for testing purposes.
+type Collector struct {
+	grpcLn     net.Listener
+	httpLn     net.Listener
+	grpcServer *grpc.Server
+	httpServer *http.Server
+	coltracepb.TraceServiceServer
+	tlsConfing *tls.Config
+	Spans      []*otlp.ScopeSpans
+}
+
+// CollectorConfig configures how a Collector should be created.
+type CollectorConfig struct {
+	TLSConfig *tls.Config
+}
+
+// NewCollector creates a new Collector based on the provided config.
+func NewCollector(cfg CollectorConfig) (*Collector, error) {
+	grpcLn, err := net.Listen("tcp4", "")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	httpLn, err := net.Listen("tcp4", "")
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	var tlsConfig *tls.Config
+	creds := insecure.NewCredentials()
+	if cfg.TLSConfig != nil {
+		creds = credentials.NewTLS(tlsConfig)
+	}
+
+	c := &Collector{
+		grpcLn:     grpcLn,
+		httpLn:     httpLn,
+		grpcServer: grpc.NewServer(grpc.Creds(creds)),
+		tlsConfing: tlsConfig,
+	}
+
+	c.httpServer = &http.Server{Handler: c, TLSConfig: tlsConfig}
+
+	coltracepb.RegisterTraceServiceServer(c.grpcServer, c)
+
+	return c, nil
+}
+
+// ClientTLSConfig returns the tls.Config that clients should
+// use to connect to the collector.
+func (c *Collector) ClientTLSConfig() *tls.Config {
```

Comment:
I was expecting for this method to return the receiver's `tlsConfig` value instead of creating a new one
Why are we creating a new instance of `tls.Config`?